system/updated/updated.py

```diff
@@ -0,0 +1,178 @@
+
+from openpilot.system.hardware.tici.agnos import flash_partition, get_target_slot_number
+
+
+
+UPDATE_DELAY = int(os.environ.get("UPDATE_DELAY", 60))
+
+API_HOST = os.getenv('API_HOST', 'https://api.commadotai.com')
+CHANNELS_API_ROOT = "v1/openpilot/channels"
+
+LOCK_FILE = os.getenv("UPDATER_LOCK_FILE", "/tmp/safe_staging_overlay.lock")
+STAGING_ROOT = os.getenv("UPDATER_STAGING_ROOT", "/data/safe_staging")
+
+CASYNC_STAGING = Path(STAGING_ROOT) / "casync"        # where the casync update is pulled
+CASYNC_TMPDIR = Path(STAGING_ROOT) / "casync_tmp"     # working directory for casync temp files
+FINALIZED = os.path.join(STAGING_ROOT, "finalized")
+
+
+def get_remote_available_channels() -> list | None:
+  try:
+    return list(requests.get(f"{API_HOST}/{CHANNELS_API_ROOT}").json())
+  except Exception:
+    cloudlog.exception("fetching remote channels")
+    return None
+
+
+def get_remote_channel_data(channel) -> tuple[BuildMetadata | None, dict | None]:
+  try:
+    data = requests.get(f"{API_HOST}/{CHANNELS_API_ROOT}/{channel}").json()
+    return build_metadata_from_dict(data["build_metadata"]), data["manifest"]
+  except Exception:
+    cloudlog.exception("fetching remote manifest failed")
+    return None, None
+
+
+def check_update_available(current_directory, other_metadata: BuildMetadata):
+  build_metadata = get_build_metadata(current_directory)
+  return build_metadata.channel != other_metadata.channel or \
+         build_metadata.openpilot.git_commit != other_metadata.openpilot.git_commit or \
+         build_metadata.openpilot.build_style != other_metadata.openpilot.build_style
+
+
+def extract_directory_helper(entry, cache_directory, directory):
+  target = casync.parse_caibx(entry["casync"]["caibx"])
+
+  cache_filename = os.path.join(CASYNC_TMPDIR, "cache.tar")
+  tmp_filename = os.path.join(CASYNC_TMPDIR, "tmp.tar")
+
+  cloudlog.info("building tarball update cache...")
+  start = time.monotonic()
+  sources = [('cache', casync.DirectoryTarChunkReader(cache_directory, cache_filename), casync.build_chunk_dict(target))]
+  cloudlog.info(f"tarball cache creation completed in {time.monotonic() - start} seconds")
+
+  sources += [('remote', casync.RemoteChunkReader(casync.get_default_store(entry["casync"]["caibx"])), casync.build_chunk_dict(target))]
+
+  cloudlog.info(f"extracting {entry['casync']['caibx']} to {directory}")
+  start = time.monotonic()
+  stats = casync.extract_directory(target, sources, directory, tmp_filename)
+  cloudlog.info(f"extraction completed in {time.monotonic() - start} seconds with {stats}")
+
+
+
+# TODO: this can be removed after all devices have moved away from overlay based git updater
+OVERLAY_MERGED = os.path.join(STAGING_ROOT, "merged")
+
+def dismount_overlay() -> None:
+  if os.path.ismount(OVERLAY_MERGED):
+    cloudlog.info("unmounting existing overlay")
+    run_cmd(["sudo", "umount", "-l", OVERLAY_MERGED])
+
+
+def setup_dirs():
+  dismount_overlay()
```

Comment:
why is there still an overlay?

Reply:
`# TODO: this can be removed after all devices have moved away from overlay based git updater`

on the first startup after the switch we need to make sure that the old overlay is removed before we can delete and cleanup the old staging root

Reply:
it won't actually be mounted, we just have to clear `/data/safe_staging` right?

Reply:
I believe I tried to delete it after a reboot and it required dismounting before I could delete it. I'll check

Reply:
repro:

```bash
sudo mount -t overlay -o lowerdir=/data/openpilot,upperdir=/data/safe_staging/upper,workdir=/data/safe_staging/metadata none /data/safe_staging/merged
rm -rf /data/safe_staging
```

`rm: cannot remove '/data/safe_staging/merged': Device or resource busy`

Reply:
ah but you are right that it won't be an issue after a reboot